### PR TITLE
Added support for JAXB2 plugins.

### PR DESCRIPTION
**Pull request by Glebcher601**
1) Added support for custom Jaxb2 plugins.
2) Made output dir generated sources seen by compiler plugin

Example:

```xml
<plugin>
        <groupId>com.github.davidmoten</groupId>
        <artifactId>xjc-maven-plugin</artifactId>
        <version>0.1.4-SNAPSHOT</version>
        <executions>
        ...
        </executions>
        <dependencies>
          <dependency>
            <groupId>org.jvnet.jaxb2_commons</groupId>
            <artifactId>jaxb2-basics</artifactId>
            <version>1.11.1</version>
          </dependency>
        </dependencies>
</plugin>
```

dependency jaxb2-basics and its dependencies will be added to classpath of jaxb-xjc execution.

### CODE REVIEW

**Comment by davidmoten** [comment]
Thanks, just to summarize we need the user to be able to add arbitrary dependencies to the classpath used for running `xjc` so we can support extensions. The way you are suggesting won't work because *xjc-maven-plugin* needs to isolate the classpath for `xjc` from the maven plugin dependencies. If you don't then bindings are still parsed but sometimes ignored (see discussion in #1). 

I started a bit of work to add the support and progress is visible in the [`support-plugins`](https://github.com/davidmoten/xjc-maven-plugin/tree/support-plugins) branch.

**Comment by Glebcher601** [comment]
So, was the problem that unwanted dependency appeared in the classpath? You said that in order for everything to work correctly xjc should be run with isolated classpath. What kind of dependencies you want to prevent from going into xjc run classpath? I need to double check it, but i suppose that taking as an example jaxb2-basics, only compile dependencies are put (jaxb2-basics-tools, jaxb2-basics-runtime, com.googlecode.javaparser). Others - such as jaxb-* are provided and won't appear, so there won't be any conflict. And one more question, do tests cover that corner case from issue #1 where bindings were ignored?

**Comment by davidmoten** [comment]
>So, was the problem that unwanted dependency appeared in the classpath? You said that in order for everything to work correctly xjc should be run with isolated classpath. What kind of dependencies you want to prevent from going into xjc run classpath?

Don't know which of the deps was the problem but one of the many many transitive dependencies of *maven-plugin-api*, *maven-core*, *maven-plugin-annotations* (just look at the deps in `xjc-maven-plugin/xjc-maven-plugin/pom.xml`).

>do tests cover that corner case from issue #1 where bindings were ignored?

Yes, the kml-2-2 generation from xsd test case in *xjc-maven-plugin-test* was affected.

**Comment by davidmoten** [comment]
I'm looking at supporting it like this (xjc dependencies are added in the configuration section):

```xml
<plugin>
        <groupId>com.github.davidmoten</groupId>
        <artifactId>xjc-maven-plugin</artifactId>
        <version>0.1.4-SNAPSHOT</version>
        <executions>
        ...
        </executions>
        <configuration>
          <dependencies>
            <dependency>
              <groupId>org.jvnet.jaxb2_commons</groupId>
              <artifactId>jaxb2-basics</artifactId>
              <version>1.11.1</version>
            </dependency>
          </dependencies>
         ...
        </configuration>
</plugin>
```

**Comment by Glebcher601** [comment]
I think you may have misunderstood me to some extent. That is exactly what i have done - allowed dependencies specified in plugin configuration(in user project which uses xjc-plugin) to occur in xjc run classpath. Thus, dependencies you stated
> one of the many many transitive dependencies of _maven-plugin-api_, _maven-core_, _maven-plugin-annotations_ (just look at the deps in `xjc-maven-plugin/xjc-maven-plugin/pom.xml`)

won't be present.

The following classpath was obtained in example I gave above:
[DEBUG] isolated classpath for call to xjc=
    /home/domainX/glebcher601/.m2/repository/org/glassfish/jaxb/jaxb-xjc/2.4.0-b180830.0438/jaxb-xjc-2.4.0-b180830.0438.jar:
    /home/domainX/glebcher601/.m2/repository/org/glassfish/jaxb/jaxb-runtime/2.4.0-b180830.0438/jaxb-runtime-2.4.0-b180830.0438.jar:
    /home/domainX/glebcher601/.m2/repository/javax/xml/bind/jaxb-api/2.4.0-b180830.0359/jaxb-api-2.4.0-b180830.0359.jar:
    /home/domainX/glebcher601/.m2/repository/javax/activation/javax.activation-api/1.2.0/javax.activation-api-1.2.0.jar:
    /home/domainX/glebcher601/.m2/repository/org/glassfish/jaxb/txw2/2.4.0-b180830.0438/txw2-2.4.0-b180830.0438.jar:
    /home/domainX/glebcher601/.m2/repository/com/sun/istack/istack-commons-runtime/3.0.7/istack-commons-runtime-3.0.7.jar:
    /home/domainX/glebcher601/.m2/repository/org/jvnet/staxex/stax-ex/1.8/stax-ex-1.8.jar:
    /home/domainX/glebcher601/.m2/repository/com/sun/xml/fastinfoset/FastInfoset/1.2.15/FastInfoset-1.2.15.jar:
    /home/domainX/glebcher601/.m2/repository/org/glassfish/jaxb/xsom/2.4.0-b180830.0438/xsom-2.4.0-b180830.0438.jar:
    /home/domainX/glebcher601/.m2/repository/com/sun/xml/bind/external/relaxng-datatype/2.4.0-b180830.0438/relaxng-datatype-2.4.0-b180830.0438.jar:
    /home/domainX/glebcher601/.m2/repository/org/glassfish/jaxb/codemodel/2.4.0-b180830.0438/codemodel-2.4.0-b180830.0438.jar:
    /home/domainX/glebcher601/.m2/repository/com/sun/xml/bind/external/rngom/2.4.0-b180830.0438/rngom-2.4.0-b180830.0438.jar:
    /home/domainX/glebcher601/.m2/repository/com/sun/xml/dtd-parser/dtd-parser/1.4/dtd-parser-1.4.jar:
    /home/domainX/glebcher601/.m2/repository/com/sun/istack/istack-commons-tools/3.0.7/istack-commons-tools-3.0.7.jar:
    /home/domainX/glebcher601/.m2/repository/org/apache/ant/ant/1.10.2/ant-1.10.2.jar:
    /home/domainX/glebcher601/.m2/repository/org/apache/ant/ant-launcher/1.10.2/ant-launcher-1.10.2.jar:
    /home/domainX/glebcher601/.m2/repository/org/jvnet/jaxb2_commons/jaxb2-basics/1.11.1/jaxb2-basics-1.11.1.jar:
    /home/domainX/glebcher601/.m2/repository/org/jvnet/jaxb2_commons/jaxb2-basics-runtime/1.11.1/jaxb2-basics-runtime-1.11.1.jar:
    /home/domainX/glebcher601/.m2/repository/org/jvnet/jaxb2_commons/jaxb2-basics-tools/1.11.1/jaxb2-basics-tools-1.11.1.jar:
    /home/domainX/glebcher601/.m2/repository/org/slf4j/slf4j-api/1.7.7/slf4j-api-1.7.7.jar:
    /home/domainX/glebcher601/.m2/repository/commons-beanutils/commons-beanutils/1.9.2/commons-beanutils-1.9.2.jar:
    /home/domainX/glebcher601/.m2/repository/commons-collections/commons-collections/3.2.1/commons-collections-3.2.1.jar:
    /home/domainX/glebcher601/.m2/repository/org/slf4j/jcl-over-slf4j/1.7.7/jcl-over-slf4j-1.7.7.jar:
    /home/domainX/glebcher601/.m2/repository/org/apache/commons/commons-lang3/3.2.1/commons-lang3-3.2.1.jar:
    /home/domainX/glebcher601/.m2/repository/com/google/code/javaparser/javaparser/1.0.11/javaparser-1.0.11.jar:
    /home/domainX/glebcher601/.m2/repository/com/github/davidmoten/xjc-maven-plugin-core/0.1.4-SNAPSHOT/xjc-maven-plugin-core-0.1.4-SNAPSHOT.jar

So I presume my solution is fine or I have misunderstood your concerns. Please let me know.

**Comment by davidmoten** [comment]
@Glebcher601 my apologies I didn't check all your changes and I see what you've done. Looks great! A unit test would be great that uses a jaxb2-plugin (and confirms that the plugin was applied). I'll review now.

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java [hunk 1]
```diff
@@ -246,10 +295,10 @@ private static String string(Artifact a) {
     private void ensureDestinationDirectoryExists() {
         for (int i = 0; i < arguments.size(); i++) {
             if (arguments.get(i).trim().equals("-d") && i < arguments.size() - 1) {
-                File dir = new File(arguments.get(i + 1));
-                if (!dir.exists()) {
-                    getLog().info("destination directory (-d option) specified and does not exist, creating: " + dir);
-                    dir.mkdirs();
+                outputDir = new File(arguments.get(i + 1));
```
Comment (davidmoten) [inline xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: we don't need a field here, just return a File from the method. The -d argument does not have to be specified. If not specified the default is the current directory which this method should return if no -d argument specified. That way we won't pass null to `buildContext.refresh`.

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java [hunk 2]
```diff
@@ -246,10 +295,10 @@ private static String string(Artifact a) {
...
+                if (!outputDir.exists()) {
+                    getLog().info("destination directory (-d option) specified and does not exist, creating: " + outputDir);
+                    outputDir.mkdirs();
```
Comment (davidmoten) [inline xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java:293]: we should return here eh, multiple -d entries will fail with xjc

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java [hunk 3]
```diff
@@ -145,6 +163,37 @@ public void execute() throws MojoExecutionException, MojoFailureException {
         return command;
     }
 
+    private Stream<String> getPluginRuntimeDependencyEntries()
```
Comment (davidmoten) [inline xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java:167]: This is great, just what I was after, thanks!

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java [hunk 4]
```diff
@@ -145,6 +163,37 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
+        Stream<Artifact> childArtifacts = artifactResolutionResults.stream()
+            .flatMap(resolutionResult -> resolutionResult.getArtifactResolutionNodes().stream())
+            .map(ResolutionNode::getArtifact);
+
+        return Stream.concat(originalArtifacts, childArtifacts)
+            .map(Artifact::getFile)
+            .map(File::getAbsolutePath);
+    }
+
+    private void addGeneratedSourcesToProjectSourceRoot() {
```
Comment (davidmoten) [inline xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: Good idea to do this. Unfortunately I'll probably still use a redundant build-helper-maven-plugin in my projects so that IDEs know that there is generated source around.

Comment (davidmoten) [reply xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: Are we assuming too much always adding to the source root? I guess we can make it configurable later if someone complains.

**Review by davidmoten — CHANGES_REQUESTED** [review]

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java (hunk 4 above)
Comment (Glebcher601) [reply xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: Ok, will revert this behavior for now

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java (hunk 3 above)
Comment (Glebcher601) [reply xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java:167]: Thanks!

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java (hunk 1 above)
Comment (Glebcher601) [reply xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: Ok, will fix it

**Comment by Glebcher601** [comment]
Thanks for review!
I'll make all necessary changes and cover new functionality with unit test.

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java [hunk 5]
```diff
@@ -292,16 +284,26 @@ private static String string(Artifact a) {
...
+                File outputDir = new File(arguments.get(i + 1));
                 if (!outputDir.exists()) {
                     getLog().info("destination directory (-d option) specified and does not exist, creating: " + outputDir);
                     outputDir.mkdirs();
+                    return outputDir;
                 }
             }
         }
+
+        final File outputDir = new File(mavenProject.getBuild().getDirectory());
```
Comment (davidmoten) [inline xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: I would expect this to be `return new File(".");`. If otherwise then the `ProcessExecutor` needs to have its current working directory set. I'm not keen on that and prefer `new File(".")` so there are no surprises as to where the output of the xjc command ends up.

**Review by davidmoten — CHANGES_REQUESTED** [review]
Thanks, looking good, just a little change with default directory

**Comment by davidmoten** [comment]
Sorry,  didn't notice those extra commits were ready for review till today (I had been looking at my email, I think it works best to add a comment as well to say ready for review again).

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java (hunk 5 above)
Comment (Glebcher601) [reply xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: All right, will fix it, however keep in mind that generated sources will then be generated in project root(which possibly clashes with VCS settings in IDE)

**Comment by Glebcher601** [comment]
Ready for review. @davidmoten please take a look.

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java [hunk 6]
```diff
@@ -296,7 +296,7 @@ private File createOutputDirectoryIfSpecifiedOrDefault() {
             }
         }
 
-        final File outputDir = new File(mavenProject.getBuild().getDirectory());
+        final File outputDir = new File(".");
```
Comment (davidmoten) [inline xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: nit: just return `new File(".")`. We don't need to ensure that the project directory exists.

**Review by davidmoten — CHANGES_REQUESTED** [review]

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java (hunk 5 above)
Comment (davidmoten) [reply xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: Yep I'm alright with that. I can add a reminder in the README.md about what happens if you leave out the `-d` parameter (and log a warning when the plugin runs). An alternative is that we add a `-d` parameter ourselves if one doesn't exist but I'd prefer not to muck with the call of `xjc` at all so the user has all the control.

File: xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java (hunk 6 above)
Comment (Glebcher601) [reply xjc-maven-plugin/src/main/java/com/github/davidmoten/xjc/XjcMojo.java]: Nice catch, will fix it

**Comment by Glebcher601** [comment]
> Ready for review. @davidmoten please take a look.

One more time

**Comment by davidmoten** [comment]
Great, thanks so much @Glebcher601! I'll clean up docs and release.

**Comment by davidmoten** [comment]
0.1.3 released to Maven Central. Should be available for download within ten minutes.

**Comment by davidmoten** [comment]
Not available for download yet for some reason.

**Comment by davidmoten** [comment]
Now it's there

**Comment by davidmoten** [comment]
@Glebcher601 just in case you're interested I've created https://github.com/davidmoten/jax-maven-plugin to handle all the tools `xjc`, `wsimport`, `wsgen`, `schemagen` using the approach we setup here.  Seems to be working fine so far.

**Comment by Glebcher601** [comment]
@davidmoten That's great! Seems like these tools will be handy(especially wsimport) for me in future to migrate some legacy stuff. Nice work!
